x-pack/legacy/plugins/lens/public/indexpattern_plugin/change_indexpattern.tsx

```diff
@@ -0,0 +1,126 @@
+import { i18n } from '@kbn/i18n';
+import { IndexPatternPrivateState, IndexPatternLayer } from './indexpattern';
+import { isLayerTransferable } from './state_helpers';
+
+export interface ChangeIndexPatternTriggerProps extends EuiButtonEmptyProps {
+  label: string;
+}
+
+export function ChangeIndexPattern({
+  indexPatterns,
+  currentIndexPatternId,
+  onChangeIndexPattern,
+  trigger,
+  layer,
+  selectableProps,
+}: {
+  trigger: ChangeIndexPatternTriggerProps;
+  indexPatterns: IndexPatternPrivateState['indexPatterns'];
+  onChangeIndexPattern: (newId: string) => void;
+  currentIndexPatternId?: string;
+  layer?: IndexPatternLayer;
+  selectableProps?: EuiSelectableProps;
+}) {
+  const [isPopoverOpen, setPopoverIsOpen] = useState(false);
+  const [selectedID, setSelectedID] = useState(
+    layer ? layer.indexPatternId : currentIndexPatternId
+  );
+
+  const indexPatternList = Object.values(indexPatterns).map(indexPattern => ({
+    ...indexPattern,
+    isTransferable: layer ? isLayerTransferable(layer, indexPattern) : undefined,
+  }));
+
+  const createTrigger = function() {
+    const { label, ...rest } = trigger;
+    return (
+      <EuiButtonEmpty
+        flush="left"
+        className="eui-textTruncate"
+        size="xs"
+        onClick={() => setPopoverIsOpen(!isPopoverOpen)}
+        {...rest}
+      >
+        {label}
+      </EuiButtonEmpty>
+    );
+  };
+
+  return (
+    <>
+      <EuiPopover
+        button={createTrigger()}
+        isOpen={isPopoverOpen}
+        closePopover={() => setPopoverIsOpen(false)}
+        className="eui-textTruncate"
+        anchorClassName="eui-textTruncate"
+        display="block"
+        panelPaddingSize="s"
+        ownFocus
+      >
+        <div style={{ minWidth: 300 }}>
+          <EuiSelectable
+            {...selectableProps}
+            searchable
+            singleSelection="always"
+            options={indexPatternList.map(indexPattern => ({
+              id: indexPattern.id,
+              label: indexPattern.title,
+              checked: indexPattern.id === selectedID ? 'on' : undefined,
+              append:
+                indexPattern && indexPattern.isTransferable !== false ? (
+                  undefined
+                ) : (
+                  <EuiIconTip
+                    color="warning"
+                    type="minusInCircle"
+                    content={i18n.translate(
+                      'xpack.lens.indexPattern.lossyIndexPatternSwitchDescription',
+                      {
+                        defaultMessage:
+                          'Not all operations are compatible with this index pattern and will be removed on switching.',
+                      }
+                    )}
+                  />
+                ),
+            }))}
+            onChange={choices => {
+              // @ts-ignore
+              const newSelectedID = choices.filter(option => option.checked)[0].id;
```

Comment:
`find` over `filter`? Also, this assumes it always finds a valid option that appears to be true but just wanted to make sure with the `!`.

```suggestion
              const newSelectedID = choices.find(({ checked }) => checked)!.id;
```

Reply:
I heard using `!` is bad practice?

Reply:
In this case `!` is the right thing IMHO - `filter` doesn't offer more safety here, if the id is not found then accessing `[0].id` will still throw. Adding a comment why the null assertion operator is justified in this instance would be nice.

Reply:
Ya with or without the `!` it will throw. The `!` just allows you to avoid typescript errors and tell typescript it will be ok and if it throws it's _my_ fault.

Reply:
K, done